frontend/public/components/monitoring/promql-expression-input.tsx

```diff
@@ -0,0 +1,395 @@
+      width: '0',
+      right: '-20px',
+      borderWidth: '10px',
+      borderStyle: 'solid',
+      borderColor: 'transparent',
+      borderLeftColor: 'var(--pf-global--palette--blue-50)',
+    },
+    marginRight: '12px',
+  },
+  '.cm-completionIcon': {
+    fontFamily: 'codicon',
+    width: '1.5em',
+    verticalAlign: 'middle',
+  },
+  '.cm-selectionMatch': {
+    backgroundColor: '#E7F1F4',
+  },
+  '.cm-completionDetail': {
+    float: 'right',
+    color: '#999',
+  },
+  '.cm-tooltip.cm-tooltip-autocomplete': {
+    boxShadow: 'var(--pf-global--BoxShadow--sm)',
+    '& > ul': {
+      fontFamily: 'var(--pf-c-code-block__pre--FontFamily), monospace',
+      fontSize: 'var(--pf-global--FontSize--sm)',
+    },
+    '& > ul > li[aria-selected]': {
+      backgroundColor: 'var(--pf-global--palette--blue-50)',
+      color: 'unset',
+    },
+    '& > ul > li': {
+      padding: '2px 1em 2px 3px',
+    },
+  },
+  '.cm-completionMatchedText': {
+    textDecoration: 'none',
+    fontWeight: 'bold',
+    color: 'var(--pf-global--palette--blue-400)',
+  },
+  '.cm-completionIcon-function, .cm-completionIcon-method': {
+    '&:after': { content: "'\\ea8c'" },
+    color: '#652d90',
+  },
+  '.cm-completionIcon-class': {
+    '&:after': { content: "'○'" },
+  },
+  '.cm-completionIcon-interface': {
+    '&:after': { content: "'◌'" },
+  },
+  '.cm-completionIcon-variable': {
+    '&:after': { content: "'𝑥'" },
+  },
+  '.cm-completionIcon-constant': {
+    '&:after': { content: "'\\eb5f'" },
+    color: '#007acc',
+  },
+  '.cm-completionIcon-type': {
+    '&:after': { content: "'𝑡'" },
+  },
+  '.cm-completionIcon-enum': {
+    '&:after': { content: "'∪'" },
+  },
+  '.cm-completionIcon-property': {
+    '&:after': { content: "'□'" },
+  },
+  '.cm-completionIcon-keyword': {
+    '&:after': { content: "'\\eb62'" },
+    color: '#616161',
+  },
+  '.cm-completionIcon-namespace': {
+    '&:after': { content: "'▢'" },
+  },
+  '.cm-completionIcon-text': {
+    '&:after': { content: "'\\ea95'" },
+    color: '#ee9d28',
+  },
+});
+
+export const selectAutocompleteOnHoverPlugin = ViewPlugin.fromClass(
+  class SelectAutocompleteOnHoverPlugin {
+    optionsLength: number = 0;
+    lastIndex: number = -1;
+
+    constructor(readonly view: EditorView) {
+      this.view.dom.addEventListener('mousemove', (this.onMouseMove = this.onMouseMove.bind(this)));
+    }
+
+    update(update: ViewUpdate) {
+      this.optionsLength = currentCompletions(update.state).length;
+    }
+
+    onMouseMove(e: Event) {
+      const element = e.target;
+      let index = -1;
+      for (
+        let dom = element as HTMLElement | null, match;
+        dom && dom !== this.view.dom;
+        dom = dom.parentNode as HTMLElement
+      ) {
+        if (
+          dom.nodeName === 'LI' &&
+          (match = /-(\d+)$/.exec(dom.id)) &&
+          +match[1] < this.optionsLength
+        ) {
+          index = +match[1];
+          break;
+        }
```

Comment:
I find this quite difficult to follow. I think it would benefit from refactoring and maybe still need comments to explain the logic. I'm not clear what is being done here, so not sure what to suggest. However, one small improvement might be to use a `while` since I think this is searching until the right node is found.

Reply:
Agreed, I refactored the code and added some comments. PTAL

Reply:
Thanks. I find this much easier to follow now 👍